src/main/java/org/folio/rest/impl/ProtectionHelper.java

```diff
@@ -0,0 +1,191 @@
+package org.folio.rest.impl;
+
+import static org.folio.invoices.utils.ErrorCodes.INVOICE_UNITS_NOT_FOUND;
+import static org.folio.invoices.utils.ErrorCodes.USER_HAS_NO_PERMISSIONS;
+import static org.folio.invoices.utils.HelperUtils.convertIdsToCqlQuery;
+import static org.folio.invoices.utils.HelperUtils.getEndpointWithQuery;
+import static org.folio.invoices.utils.HelperUtils.handleGetRequest;
+import static org.folio.invoices.utils.ResourcePathResolver.ACQUISITIONS_MEMBERSHIPS;
+import static org.folio.invoices.utils.ResourcePathResolver.ACQUISITIONS_UNITS;
+import static org.folio.invoices.utils.ResourcePathResolver.resourcesPath;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import me.escoffier.vertx.completablefuture.VertxCompletableFuture;
+import one.util.streamex.StreamEx;
+import org.apache.commons.collections4.CollectionUtils;
+import org.folio.HttpStatus;
+import org.folio.invoices.rest.exceptions.HttpException;
+import org.folio.invoices.utils.HelperUtils;
+import org.folio.invoices.utils.ProtectedOperationType;
+import org.folio.rest.acq.model.units.AcquisitionsUnit;
+import org.folio.rest.acq.model.units.AcquisitionsUnitCollection;
+import org.folio.rest.acq.model.units.AcquisitionsUnitMembership;
+import org.folio.rest.acq.model.units.AcquisitionsUnitMembershipCollection;
+import org.folio.rest.tools.client.interfaces.HttpClientInterface;
+
+import io.vertx.core.Context;
+
+public class ProtectionHelper extends AbstractHelper {
+
+  public static final String ACQUISITIONS_UNIT_ID = "acquisitionsUnitId";
+  public static final String ACQUISITIONS_UNIT_IDS = "acqUnitIds";
+  public static final String NO_ACQ_UNIT_ASSIGNED_CQL = "cql.allRecords=1 not %s <> []";
+  static final String GET_UNITS_BY_QUERY = resourcesPath(ACQUISITIONS_UNITS) + SEARCH_PARAMS;
+  static final String GET_UNITS_MEMBERSHIPS_BY_QUERY = resourcesPath(ACQUISITIONS_MEMBERSHIPS) + SEARCH_PARAMS;
+
+  public ProtectionHelper(HttpClientInterface httpClient, Map<String, String> okapiHeaders, Context ctx, String lang) {
+    super(httpClient, okapiHeaders, ctx, lang);
+}
+
+  /**
+   * This method determines status of operation restriction based on unit IDs from {@link CompositePurchaseOrder}.
+   * @param unitIds list of unit IDs.
+   *
+   * @throws HttpException if user hasn't permissions or units not found
+   */
+  public CompletableFuture<Void> isOperationRestricted(List<String> unitIds, ProtectedOperationType operation) {
+    if (CollectionUtils.isNotEmpty(unitIds)) {
+      return getUnitsByIds(unitIds)
+        .thenCompose(units -> {
+          if (unitIds.size() == units.size()) {
+            if (applyMergingStrategy(units, operation)) {
+              return verifyUserIsMemberOfOrdersUnits(unitIds);
```

Comment:
This operation is associated to invoice so method name should be something like `verifyUserIsMemberOfInvoiceUnits`?

Reply:
Renamed to `verifyUserIsMemberOfAcqUnits` because the helper is going to be used for invoice and voucher records